<a href="https://colab.research.google.com/github/Harshita0201/Machine_Learning/blob/main/Breast_Cancer_Detection_Logistic_Regression).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Logistic Regression

## Importing the libraries

In [1]:
import pandas as pd

## Importing the dataset

In [2]:
dataset = pd.read_csv('breast_cancer.csv')
X = dataset.iloc[: , 1:-1].values #independent variables from all the cols except the last one, we also ignore the first col as it doest not contribute to prediction 
y = dataset.iloc[:, -1].values #dependent variable or the class(2 for benign, 4 for malignant) in last col of dataset

## Splitting the dataset into the Training set and Test set

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0) 
#137 observations in test set and rest 546 in training set

## Training the Logistic Regression model on the Training set

In [5]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0) #create instance of class
classifier.fit(X_train, y_train) #train it on or training set

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

## Predicting the Test set results

In [6]:
#predicting it for new observation 
y_pred = classifier.predict(X_test)

## Making the Confusion Matrix

In [7]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred) #ground truth vs the predicted values
print(cm)

[[84  3]
 [ 3 47]]


84 correct predictions that the tumor is benign and 47 correct predictions that the tumor is milagnant, 3 incorrect predictions that tumor is benign  (false positive) and 3 incorrect predictions that tumor is milagnant i.e (false negative).
we have 84+47=131 correct predictions out of 137 observations from test set.

In [8]:
#accuracy :
#no. of predicted corrext / total observations
#(84+47)/(84+47+3+3)

0.9562043795620438

## Computing the accuracy with k-Fold Cross Validation

In [9]:
# K-Fold Cross Validation: A technique where we build 10 train-test folds
# and hence ending up with 10 accuracies and then we will include average of these 10 accuracies to get a relevant measure of our LogisticRegression model
# We implement this using the cross_val_score class under the model_selection module of sklearn

from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator=classifier, X = X_train, y = y_train, cv=10) #the argument estimator linked the accuracy function to our claassifier, and X and y takes the values of matrix of features on which we apply this accuracy function cv depicts the number of folds we want, here 10 
#accuracies have list of 10 accuracy of the training set folds
print("Accuracy: {:.2f} %".format(accuracies.mean()*100)) #print the average of all accuracies using the mean method of cross_val_score class
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100)) # to know the varaince in the accuracies, using the std method of cross_val_score class, less the variance good is the prediction

Accuracy: 96.70 %
Standard Deviation: 1.97 %


This means that 10 accuracies fall in the 96.70-1.97% to 96.70+1.97% accuracy or (94-98%)